mne/decoding/time_delaying_ridge.py

```diff
@@ -0,0 +1,186 @@
+    len_trf = (ac.shape[2] + 1) // 2
+    n_ch = ac.shape[0]
+    xxt = np.zeros([n_ch * len_trf] * 2)
+    for ch0 in range(n_ch):
+        for ch1 in range(n_ch):
+            xxt_temp = np.zeros((len_trf, len_trf))
+            xxt_temp[0, :] = ac[ch0, ch1, len_trf - 1:]
+            xxt_temp[:, 0] = ac[ch0, ch1, len_trf - 1::-1]
+            for ii in range(1, len_trf):
+                xxt_temp[ii, ii:] = ac[ch0, ch1, len_trf - 1:-ii]
+                xxt_temp[ii:, ii] = ac[ch0, ch1, len_trf - 1:ii - 1:-1]
+            xxt[ch0 * len_trf:(ch0 + 1) * len_trf,
+                ch1 * len_trf:(ch1 + 1) * len_trf] = xxt_temp
+    return xxt
+
+
+class TimeDelayingRidge(BaseEstimator):
+    """Ridge regression of data with time delays.
+
+    Parameters
+    ----------
+    smin : int
+        Minimum sample number.
+    smax : int
+        Maximum sample number.
+    alpha : float
+        The ridge (or laplacian) regularization.
+    reg_type : str
+        Can be "ridge" (default) or "laplacian".
+    fit_intercept : bool
+        If True (default), the sample mean is removed before fitting.
+    """
+
+    def __init__(self, smin, smax, alpha=0., reg_type='ridge',
+                 fit_intercept=True):  # noqa: D102
+        assert smin <= smax
+        self._smin = -smax
+        self._smax = -smin + 1
+        self._alpha = float(alpha)
+        self._reg_type = reg_type
+        self._fit_intercept = fit_intercept
+
+    def fit(self, X, y):
+        """Estimate the coefficients of the linear model.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples, n_features)
+            The training input samples to estimate the linear coefficients.
+        y : array, shape (n_samples, n_outputs)
+            The target values.
+
+        Returns
+        -------
+        self : instance of LinearModel
+            Returns the modified instance.
+        """
+        # These are split into two functions because it's possible that we
+        # might want to allow people to do them separately (e.g., to test
+        # different regularization parameters).
+        x_xt, x_y, n_ch_x = _compute_corrs(X, y, self._smin, self._smax,
+                                           self._fit_intercept)
+        self.coef_ = _fit_corrs(x_xt, x_y, n_ch_x,
+                                self._reg_type, self._alpha, n_ch_x)
+        self.coef_ = self.coef_[..., ::-1]
+        return self
+
+    def predict(self, X):
+        """Predict the output.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples, n_features)
+            The data.
+
+        Returns
+        -------
+        X : ndarray
+            The predicted response.
+        """
+        out = np.zeros((X.shape[0], self.coef_.shape[0]))
+        offset = max(self._smin, 0)
+        for oi in range(self.coef_.shape[0]):
+            for fi in range(self.coef_.shape[1]):
+                temp = np.convolve(X[:, fi], self.coef_[oi, fi][::-1])
```

Comment:
ahhhh convolution! https://www.youtube.com/watch?v=92gP2J0CUjc&feature=youtu.be&t=1m32s